src/fluent/watch.ts

```diff
@@ -31,64 +73,134 @@ export async function ExecWatch<T extends GenericClass>(
+        }
+      }
+    };
+
+    const cleanup = () => {
+      if (!doneCalled) {
+        doneCalled = true;
+        stream.removeAllListeners();
+      }
+    };
+
+    try {
+      // Make the actual request
+      const response = await fetch(url, { ...opts });
+
+      // If the request is successful, start listening for events
+      if (response.ok) {
+        const { body } = response;
+
+        // Reset the retry count
+        retryCount = 0;
+
+        stream.on("error", onError);
+        stream.on("close", cleanup);
+        stream.on("finish", cleanup);
+
+        // Listen for events and call the callback function
+        stream.on("data", line => {
+          try {
+            // Parse the event payload
+            const { object: payload, type: phase } = JSON.parse(line) as {
+              type: WatchPhase;
+              object: InstanceType<T>;
+            };
+
+            // Call the callback function with the parsed payload
+            void callback(payload, phase as WatchPhase);
+          } catch (err) {
+            watchCfg.logFn?.(err, "watch callback error");
+          }
+        });
+
+        body.on("error", onError);
+        body.on("close", cleanup);
+        body.on("finish", cleanup);
+
+        // Pipe the response body to the stream
+        body.pipe(stream);
+      } else {
+        throw new Error(`watch failed: ${response.status} ${response.statusText}`);
+      }
+    } catch (e) {
+      onError(e);
     }
-  };
-
-  try {
-    // Make the actual request
-    const response = await fetch(url, opts);
-
-    // If the request is successful, start listening for events
-    if (response.ok) {
-      const { body } = response;
-
-      // Bind connection events to the close function
-      body.on("error", close);
-      body.on("close", close);
-      body.on("finish", close);
-
-      // Create a readline interface to parse the stream
-      const rl = readline.createInterface({
-        input: response.body!,
-        terminal: false,
-      });
-
-      // Listen for events and call the callback function
-      rl.on("line", line => {
-        try {
-          // Parse the event payload
-          const { object: payload, type: phase } = JSON.parse(line) as {
-            type: WatchPhase;
-            object: InstanceType<T>;
-          };
-
-          // Call the callback function with the parsed payload
-          void callback(payload, phase as WatchPhase);
-        } catch (ignore) {
-          // ignore parse errors
+
+    // On unhandled errors, retry the watch
+    async function reload(e: Error) {
+      // If there are more attempts, retry the watch
+      if (watchCfg.retryMax! > retryCount) {
+        retryCount++;
+
+        watchCfg.logFn?.(`retrying watch ${retryCount}/${watchCfg.retryMax}`);
+
+        // Sleep for the specified delay or 5 seconds
+        await new Promise(r => setTimeout(r, (watchCfg.retryDelaySec ?? 5) * 1000));
```

Comment:
Do you want to assign the default value in the same way retryMax is so you are not doing inline assignments?

Reply:
Good point, I'll update that